inference_overview.md

```diff
@@ -0,0 +1,449 @@
+
+This method works better for smaller and simpler models.
+
+- Complex models often require large frameworks or libraries, and other
+  initialization, which may incur a higher cost to run inside the JavaScript
+  sandbox.
+- Not all model code can be translated into Wasm or JavaScript in a
+  straightforward way. It may be more difficult to translate complex models and
+  supporting code. There is a tradeoff between model size and memory cost.
+
+### Inference with externally provided models
+
+In this method, we will support loading binary model data from external sources.
+The models will be fetched periodically from a linked source (e.g. [Google Cloud
+Storage buckets][13], [Amazon S3 buckets][14]) and loaded into memory. An API
+will be provided to `generateBid()` to run inference against the loaded models.
+The whole operation of invoking `runInference()`, running predictions using the
+loaded models and returning the values to the caller will take place inside a
+single TEE. Data will not leave the TEE, and will not be observable externally
+either. Outside observers will not be directly able to observe how many times
+inference was invoked, if at all.
+
+We will initially support [Tensorflow][15] and [PyTorch][16] based models.
+
+We will provide the following APIs, which will be usable from ad tech code:
+
+- **Run inference using the external models**: This is the main way to use
+  inference. A batch of inputs can be constructed in ad tech code and passed to
+  the API. Inference will run for each input on the given pre-loaded model and
+  the responses will be returned. The call, for illustrative purposes, may look
+  like the following:
+- **Get the list of paths of models available for use**: This would get the list
+  of models currently loaded and available for use in the server. The paths of
+  models would typically encapsulate version information (see [Versions in
+  B&A][17]). Along with the versioning strategy, this API gives the ad tech
+  precise control in choosing versions of models to run inference against. The
+  call, for illustrative purposes, may look like the following:
+
+Note that these APIs are for clarity only and may not reflect the actual
+specification. We will define the actual API in more detail in future iterations
+of this explainer. These APIs will also be callable from Wasm.
+
+## Factorized and Unified models
+
+Model inputs can use private features (e.g. user data) and non-private features
+(e.g. ad data). Private features are available only inside the TEE and so, such
+inference must run inside the TEE. This is because the private data should not
+leave the privacy preserving boundary. Non-private features can be available on
+the contextual path ([see B&A system design][18]), and models relying only on
+such data can run outside the TEE, on ad tech hardware, and may also pre-compute
+inferences.
+
+Common bid prediction models use contextual, ad and user features for
```

Comment:
Nit. It sounds like "ad and user" is one feature.

Common bid prediction models utilize contextual features, ad features, and user features to generate accurate predictions.

Reply:
Ack. Reading the next sentence should clarify any confusion, if it was there.